test_data/asm/book/declarations.asm

```diff
@@ -1,12 +1,17 @@
 mod utils {
     // This defines a function by means of a lambda expression that
-    // computes the sum of an array of values.
-    let sum = |len, arr| match len { 0 => 0, _ => arr[len - 1] + sum(len - 1, arr) };
-    // A simple function that returns the input incremented by one.
-    let incremented = |x| x + 1;
+    // computes the sum of an array of values. We fully specify its type.
+    let sum: int, int[] -> int = |len, arr| match len {
+        0 => 0,
+        _ => arr[len - 1] + sum(len - 1, arr)
+    };
+    // A simple function that returns the input incremented by one,
+    // as an expression.
+    let incremented: expr -> expr = |x| x + 1;
     // This is a function that takes an expression as input and returns
     // a constraint enforcing this expression increments by a certain value
     // between rows.
+    // The type will be inferred here because `'` is only valid on `expr`.
```

Comment:
In principle yes, the `next` operator is "linear" in that `(5*x + y + 1)' = 5*(x') + y' + 1`, but I think the condenser would have to do that transformation and it currently does not do it (although I think it would be simple), but I might be wrong.